src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -513,6 +537,193 @@ internal void ReadToken(string encodedJson)
+            encodedJsonSpanCopy = encodedJsonSpanCopy.Slice(dotIndex + 1);
+            dotIndex = encodedJsonSpanCopy.IndexOf("."); // second dot
+            if (dotIndex == -1)
+                throw LogHelper.LogExceptionMessage(new SecurityTokenMalformedException(LogMessages.IDX14120));
+
+            // segment between first and second dot, can be a payload for JWS or an encrypted key for JWE (b)
+            ReadOnlySpan<char> secondSpan = encodedJsonSpanCopy.Slice(0, dotIndex);
+            Dot2 = Dot1 + dotIndex + 1;
+
+            if (dotIndex == encodedJsonSpanCopy.Length - 1)
+            {
+                dotIndex = -1;
+            }
+            else
+            {
+                // c.d.e
+                encodedJsonSpanCopy = encodedJsonSpanCopy.Slice(dotIndex + 1);
+                dotIndex = encodedJsonSpanCopy.IndexOf('.'); // third dot
+            }
+
+            if (dotIndex == -1)
+            {
+                // JWS has two dots
+                // JWS: https://www.rfc-editor.org/rfc/rfc7515
+                // Format: https://www.rfc-editor.org/rfc/rfc7515#page-7
+
+                // add length of header and second span, plus 2 dots
+                IsSigned = !(headerSpan.Length + secondSpan.Length + 2 == encodedJsonSpan.Length);
+                try
+                {
+                    Header = CreateClaimSet(headerSpan.ToString(), 0, headerSpan.Length, CreateHeaderClaimSet); // CreateClaimSet invokes Base64UrlEncoding.Decode which accepts only a string
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(
+                        LogMessages.IDX14102,
+                        LogHelper.MarkAsUnsafeSecurityArtifact(headerSpan.ToString(), t => t.ToString())), // TODO: Add an overload to LogHelper.MarkAsUnsafeSecurityArtifact that accepts span?
+                        ex));
+                }
+
+                try
+                {
+                    Payload = CreateClaimSet(secondSpan.ToString(), 0, secondSpan.Length, CreatePayloadClaimSet);
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(
+                        LogMessages.IDX14101,
+                        LogHelper.MarkAsUnsafeSecurityArtifact(secondSpan.ToString(), t => t.ToString())),
+                        ex));
+                }
+            }
+            else
+            {
+                // JWE: https://www.rfc-editor.org/rfc/rfc7516
+                // Format: https://www.rfc-editor.org/rfc/rfc7516#page-8
+                // empty payload for JWE's {encrypted tokens}.
+                Payload = new JsonClaimSet();
+
+                if (dotIndex == encodedJsonSpanCopy.Length) // TODO: Should this be encodedJsonSpanCopy.Length - 1? Using encodedJsonSpanCopy.Length doesn't add up
```

Comment:
>Length [](http://example.com/codeflow?start=52&length=6)

what's the issue with .Length? 'dotIndex' is the first dot character, so I'd think `dotIndex == encodedJsonSpanCopy.Length` would work.

Reply:
My thinking is that the value of EncodedJsonSpanCopy.Length will fall outside the range of the string and EncodedJsonSpanCopy.Length - 1 would be more appropriate.